File: src/lexer/lexer.rs
// use anyhow::Result;

#[derive(Debug, PartialEq)]
pub enum Token {
    Identifier,
    Package,
    Import(String),
    Class(String),
    Public(String),
    Protected(String),
    Private(String),
    Static(String),
    Abstract(String),
    Final(String),
    LBracket,
    RBracket,
    Semicolon, // ;
    Colon,     // :
    Eq,
    Assign,
    LParen,
    RParen,
    Byte(String),
    Short(String),
    Int(String),
    Long(String),
    Char(String),
    Float(String),
    Double(String),
    Boolean(String),
    Void(String),
    If(String),
}

pub struct Lexer {
    position: usize,
    read_position: usize,
    char: u8,
    input: Vec<u8>,
}

impl Lexer {
    pub fn new(input: String) -> Lexer {
        return Lexer {
            position: 0,
            read_position: 0,
            char: 0,
            input: input.into_bytes(),
        };
    }

    // fn next_token(&mut self) -> Result<Token> {
    pub fn next_token(&mut self) {
        while self.read_position < self.input.len() {
            let c = self.current_char();

            println!("{:?}", c as char);
            self.read_position += 1;
        }

        // let token = match self.char {
        //     b'{' => Token::LBracket,
        //     b'}' => Token::RBracket,
        //     b';' => Token::Semicolon,
        //     b':' => Token::Colon,
        //     b'=' => Token::Assign,
        //     b'(' => Token::LParen,
        //     b')' => Token::RParen,
        //     _ => todo!("Precisa ver isso aí"),
        // };
        // return Ok(token);
    }

    fn current_char(&mut self) -> u8 {
        if self.read_position >= self.input.len() {
            return 0;
        } else {
            return self.input[self.read_position];
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use anyhow::Result;

    #[test]
    fn get_next_token() -> Result<()> {
        let input = r#"
          package example;
          import java.util.ArrayList;
          import java.util.List;

          public class Example {
            private ArrayList<String> names;

            public Example() {
              names = new ArrayList<>();
            }

            public void addName(String name) {
              names.add(name);
            }

            public List<String> getNames() {
              return new ArrayList<>(names);
            }
          }
          "#;
        let lex = Lexer::new(input.into());
        // let tokens: Vec<Token> = vec![Token::Package, Token::Identifier];

        return Ok(());
    }
}

File: lexer.ts
import { match } from "ts-pattern";
const Keywords = [];
enum TokenKind {
  Identifier = "Identifier",
  Package = "Package",
  Import = "Import",
  Class = "Class",
  Public = "Public",
  Protected = "Protected",
  Private = "Private",
  Static = "Static",
  Abstract = "Abstract",
  Final = "Final",
  LBracket = "LBracket",
  RBracket = "RBracket",
  Semicolon = "Semicolon", // ;
  Colon = "Colon", // :
  Eq = "Eq",
  Assign = "Assign",
  GreaterThan = "GreaterThan",
  LowerThan = "LowerThan",
  LParen = "LParen",
  RParen = "RParen",
  Byte = "Byte",
  Short = "Short",
  Int = "Int",
  Long = "Long",
  Char = "Char",
  Float = "Float",
  Double = "Double",
  Boolean = "Boolean",
  Void = "Void",
  If = "If",
}
type Token = {
  tokenType: TokenKind;
  identifier: string;
};
export class Lexer {
  private position: number = 0;
  private readPosition: number = 0;
  private ch!: string;
  constructor(private input: string) {
    this.nextChar();
  }
  nextToken() {
    this.skipWhitespace();
    let token: Token | null = null;
    match(this.ch)
      .with("{", () => (token = this.newToken(TokenKind.LBracket, this.ch)))
      .with("}", () => (token = this.newToken(TokenKind.RBracket, this.ch)))
      .with(";", () => (token = this.newToken(TokenKind.Semicolon, this.ch)))
      .with(":", () => (token = this.newToken(TokenKind.Colon, this.ch)))
      .with("(", () => (token = this.newToken(TokenKind.LParen, this.ch)))
      .with(")", () => (token = this.newToken(TokenKind.RParen, this.ch)))
      .with("=", () => {
        if (this.peekChar() === "=") {
          token = this.newToken(TokenKind.Eq, "==");
        } else {
          token = this.newToken(TokenKind.Assign, this.ch);
        }
      })
      .with(">", () => (token = this.newToken(TokenKind.GreaterThan, this.ch)))
      .with("<", () => (token = this.newToken(TokenKind.LowerThan, this.ch)));

    if (this.isAlphabetic(this.ch)) {
      const identifier = this.readIdentifier();
      return this.newToken(TokenKind.Identifier, identifier);
    } else if (this.isNumber(this.ch)) {
      const num = this.readNumber();
      return this.newToken(TokenKind.Int, num);
    }
    this.nextChar();
    return token;
  }

  private newToken(tokenType: TokenKind, identifier: string): Token {
    return { tokenType, identifier };
  }
  private skipWhitespace() {
    do {
      this.nextChar();
    } while (
      this.ch === " " ||
      this.ch === "\n" ||
      this.ch === "\t" ||
      this.ch === "\r"
    );
  }

  private peekChar(): string {
    if (this.readPosition >= this.input.length) {
      return "\0";
    }
    return this.input[this.readPosition];
  }

  isNumber(ch: string): boolean {
    const char = ch.charCodeAt(0);
    return "0".charCodeAt(0) <= char && char <= "9".charCodeAt(0);
  }

  isAlphabetic(ch: string): boolean {
    const char = ch.charCodeAt(0);
    return (
      ("a".charCodeAt(0) <= char && "z".charCodeAt(0) >= char) ||
      ("A".charCodeAt(0) <= char && "Z".charCodeAt(0) >= char) ||
      char === "_".charCodeAt(0)
    );
  }

  private nextChar() {
    if (this.readPosition >= this.input.length) {
      this.ch = "\0";
    } else {
      this.ch = this.input[this.readPosition];
    }
    this.position = this.readPosition;
    this.readPosition += 1;
  }

  private readIdentifier(): string {
    while (this.isAlphabetic(this.ch)) {
      this.nextChar();
    }
    return this.input.substring(this.position, this.readPosition);
  }

  private readNumber(): string {
    while (this.isNumber(this.ch)) {
      this.nextChar();
    }
    return this.input.substring(this.position, this.readPosition);
  }
}
